### PR TITLE
Support InfluxDB 2.0

### PR DESCRIPTION
**Pull request by ChrisBr**
Initial support for InfluxDB > 1.8.

## Client config

Some of the client config changed

```ruby
InfluxDB::Rails.configure do |config|
  ## The only settings you actually need to set are org, bucket and token.
  config.client.bucket = "bucket"
  config.client.org = "org"
  config.client.token = "token"
  # config.client.url = "localhost:8086",

  ## Various other client and connection options. These are used to create
  ## an `InfluxDB2::Client` instance (i.e. `InfluxDB::Rails.client`).
  ##
  ## See docs for the influxdb-client gem for the canonical list of options:
  ## https://github.com/influxdata/influxdb-client-ruby#creating-a-client
  ##
  # config.client.use_ssql = true
  # config.client.open_timeout = 5.seconds
  # config.client.write_timeout = 5.seconds
  # config.client.read_timeout = 60.seconds
  # config.client.time_precisions = InfluxDB2::WritePrecision::MILLISECOND
  # config.client.retries = 0
  # config.client.max_retry_delay_ms = 10_000
  # config.client.async = true

  ## Disable rails framework hooks.
  # config.ignored_hooks = ['sql.active_record', 'render_template.action_view']

  # Modify tags on the fly.
  # config.tags_middleware = ->(tags) { tags }

  ## Set the application name to something meaningful, by default we
  ## infer the app name from the Rails.application class name.
  # config.application_name = Rails.application.class.parent_name
end
```

## Fields vs Values

I used this opportunity to rename values to fields because this is how they're called in InfluxDB so we avoid any confusion and use the official upstream terms.

## Logging

Made it possible to provide a logger only for InfluxDB.

## Empty tags and values

The [client library](https://github.com/influxdata/influxdb-client-ruby/blob/master/lib/influxdb2/client/point.rb#L138-L164) now filters out empty values so I removed this code here.

### CODE REVIEW

**Comment by ChrisBr** [comment]
No idea why the tests fail, they work locally for me.

Also there is an issue with the timestamps https://github.com/influxdata/influxdb-client-ruby/issues/68

**Comment by hennevogel** [comment]
> No idea why the tests fail, they work locally for me.

Go find out :wink:

File: lib/influxdb/rails/middleware/request_subscriber.rb [hunk 1]
```diff
@@ -28,16 +28,9 @@ def values
             controller: duration,
             view:       (payload[:view_runtime] || 0).ceil,
             db:         (payload[:db_runtime] || 0).ceil,
-            started:    started,
+            started:    start.utc,
           }
         end
-
-        def started
-          InfluxDB.convert_timestamp(
```
Comment (hennevogel) [inline lib/influxdb/rails/middleware/request_subscriber.rb]: I'm supposed to take care of this myself now?

**Review by hennevogel — CHANGES_REQUESTED** [review]

**Comment by hennevogel** [comment]
> InfluxDB::Rails.client = InfluxDB2::Client.new()

Hmmm. Not sure it's fair to leave people completely alone with this. I mean, unlike us, they come from the other direction you know? Just looking for some instrumentation...

**Comment by ChrisBr** [comment]
> Hmmm. Not sure it's fair to leave people completely alone with this. I mean, unlike us, they come from the other direction you know? Just looking for some instrumentation...

Okay, makes sense. We have to update the config then though.

File: lib/influxdb/rails/middleware/request_subscriber.rb (hunk 1 above)
Comment (ChrisBr) [reply lib/influxdb/rails/middleware/request_subscriber.rb]: No the influxdb-client-ruby gem does it now. You have to pass in the precision when creating the client.

We either have to do a mapping from our symbols or let people set these constants here too.

```ruby
client = InfluxDB2::Client.new('https://localhost:8086', 'my-token',
  bucket: 'my-bucket',
  org: 'my-org',
  precision: InfluxDB2::WritePrecision::NANOSECOND)
```

https://github.com/influxdata/influxdb-client-ruby#creating-a-client

**Comment by ChrisBr** [comment]
@hennevogel do you want to create a new branch from the current master like 1.0-stable and develop 2.0 on master?

File: lib/influxdb/rails/middleware/request_subscriber.rb (hunk 1 above)
Comment (hennevogel) [reply lib/influxdb/rails/middleware/request_subscriber.rb]: So the client converts internally then? cool

**Comment by hennevogel** [comment]
> do you want to create a new branch from the current master like 1.0-stable and develop 2.0 on master?

Yeah, sounds like a plan.

**Comment by hennevogel** [comment]
> Okay, makes sense. We have to update the config then though.

Well maybe it's enough to provide sensible defaults in the generator?

File: README.md [hunk 2]
```diff
@@ -242,7 +243,7 @@ class ApplicationController
 
   def set_influx_data
     InfluxDB::Rails.current.tags = { user: current_user.id }
-    InfluxDB::Rails.current.values = { redis_value: redis_value }
+    InfluxDB::Rails.current.fields = { redis_value: redis_value }
```
Comment (ChrisBr) [inline README.md:238]: Should we make this backward compatible or break because it's a major update?

File: lib/influxdb-rails.rb [hunk 3]
```diff
@@ -21,41 +22,53 @@ module InfluxDB
...
-            retry:          cfg.retry,
-            open_timeout:   cfg.open_timeout,
-            read_timeout:   cfg.read_timeout,
-            max_delay:      cfg.max_delay,
-            time_precision: cfg.time_precision
+          InfluxDB2::Client.new(
+            cfg.url,
+            cfg.token,
+            org:           cfg.org,
+            bucket:        cfg.bucket,
```
Comment (ChrisBr) [inline lib/influxdb-rails.rb:51]: This will be a breaking change.

File: lib/influxdb-rails.rb [hunk 4]
```diff
@@ -21,41 +22,53 @@ module InfluxDB
...
+            precision:     cfg.precision,
```
Comment (ChrisBr) [inline lib/influxdb-rails.rb:52]: Should we keep `time_precision` or follow the upstream config?

File: lib/influxdb/rails/configuration.rb [hunk 5]
```diff
@@ -28,25 +28,37 @@ def load_defaults
...
       )
 
       def initialize
         load_defaults
       end
+
+      def write_options
+        InfluxDB2::WriteOptions.new(
+          write_type:  WRITE_TYPE[async],
+          max_retries: retries
```
Comment (ChrisBr) [inline lib/influxdb/rails/configuration.rb]: Should we follow upstream or keep the same config?

File: lib/influxdb/rails/configuration.rb [hunk 6]
```diff
@@ -28,25 +28,36 @@ def load_defaults
...
       )
 
       def initialize
         load_defaults
       end
+
+      def write_options
+        InfluxDB2::WriteOptions.new(
+          write_type:  WRITE_TYPE[async],
+          max_retries: max_retries
```
Comment (ChrisBr) [inline lib/influxdb/rails/configuration.rb:58]: Should we follow upstream configuration or keep `retries`? I also find `max_retries` more descriptive.

**Comment by antonengelhardt** [comment]
@ChrisBr Are you planning to complete work on this PR? We'd really love to have support for Influx > 2.0, preferably 2.7. Thanks!

**Comment by ChrisBr** [comment]
Not actively working on it but happy to review a pull request!

**Comment by araslanov-e** [comment]
@hennevogel please add support for 2.0

File: README.md (hunk 2 above)
Comment (hennevogel) [reply README.md:238]: There isn't much backward compatibility between Influx 1.x and 2.x, I don't think we should be the exception.

File: lib/influxdb-rails.rb (hunk 4 above)
Comment (hennevogel) [reply lib/influxdb-rails.rb:52]: follow upstream

File: lib/influxdb/rails/configuration.rb (hunk 5 above)
Comment (hennevogel) [reply lib/influxdb/rails/configuration.rb]: follow upstream

File: lib/influxdb/rails/configuration.rb (hunk 6 above)
Comment (hennevogel) [reply lib/influxdb/rails/configuration.rb:58]: follow upstream

File: lib/influxdb-rails.rb (hunk 3 above)
Comment (hennevogel) [reply lib/influxdb-rails.rb:51]: This is InfluxDB 2.x!

BREAK AND DANCE

**Review by hennevogel — APPROVED** [review]
Here we go!

**Comment by hennevogel** [comment]
Now on to the documentation and update guide.

**Comment by antonengelhardt** [comment]
Thanks a lot! @ChrisBr @hennevogel

**Pull request merged**